discovery.md

```diff
@@ -650,30 +630,49 @@ upon return of the `202 Accepted` from the original request.
+the discovery endpoint MUST assign an appropriate value to the omitted
+field(s).  An appropriate `id` MUST be globally unique.  An appropriate `epoch`
+MUST be greater than the any existing `epoch` for a service with the same `id`.
+Whether an `id` is given or supplied, any existing Service definition with that
+`id` MUST be replaced by the Service definition in the request and any assigned
+values. If an `epoch` is given in the request it MUST be greater than the
+`epoch` of any existing Service of the same `id` or the request MUST fail.
+
+If the Discovery Endpoint is unable to successully add every Service in the
+incoming request then an error MUST be generated and none of the specified
+Services are to be added or updated in the Discovery Endpoint.
+
+The Services in the request MUST be processed in the order in which they are
+given. While any error during the update MUST cause the rollback of any effect
+of the request, the processing of each Service in the request MUST be
+applied as if all previous Services in the request were already completed.
+This means that attribute uniqueness checking, such as on the `name`
+attribute, MUST be done on the new state of all Services as if all previously
+specified Services in the incoming request were already successfully processed.
 
-If any Service specified in the request matches the `name` of any existing
-Service, or the same `name` is used more than once within the request, then
-an error MUST be generated.
+For example, if a Discovery Endpoint has a Service called `MyService`, it is
+possible to rename that Service to `YourService` and then to add, or update,
+another Service to use `MyService` as its `name`. But this is only true if
+the rename of the first Service happens first the request message.
 
-If the Discovery Endpoint is unable to successully add all of the Services
-in the incoming request then an error MUST be generated and none of the
-specified Services are to be added to the Discovery Endpoint.
+Likewise, while it might be less than optimal, it is technically possible for
+a request to update a Service multiple times if the same `id` is used more
+than once within a request.
 
 The follow responses are defined by this specification:
-- `201 Created` if all the specified Services were processed successfully.
-  - If there was only one Service in the incoming request, then the response
-    MUST include an HTTP `Location` Header with a value that points to the
-    correspdonding Service.
-  - If the request contained more than one Service, then the HTTP `Location`
-    Header MUST NOT appear in the response.
-  - The HTTP Response Body MUST include an array of `id` values whose order
-    and values MUST match the list of Services in the request.
+- `200 OK` if all the specified Services were processed successfully
+  - The HTTP Response Body MUST include an array of `id` and `epoch` values
+    whose order and values MUST match the list of Services in the request
```

Comment:
Perhaps a missing sophistication is that `id` and `epoch` would be included for newly created Services (i.e. those without a preset `id`.  For those which were replacements the old values could be returned.  This would be useful and consistent with the DELETE behavior but imply increased implementation work.

Reply:
hmmm at first I didn't see the value in returning the old service (and on delete too), but your comments reminded me that I think I've seen some service (like github's events I think) that will return the old data (at least for some fields) so you can do a diff if you really wanted. And I have found that useful.

Let's see what other people think - I can go either way right now

Reply:
As per the call, both POST and DELETE will (in case of success) return the entire Service value resulting from / just prior to (respectively) the successful processing of the request.